File: day4/second/main.py
card_counts = []

with open("../input.txt") as file:
    card_idx = 0
    while (line := file.readline()):
        parts = line.split("|")
        winners = sorted(map(int, parts[0][parts[0].index(":")+1:].strip().split()))
        nums = sorted(map(int, parts[1].strip().split()))

        win_count = 0
        i, j = 0, 0
        while i < len(winners) and j < len(nums):
            if winners[i] == nums[j]:
                win_count += 1
                i += 1
                j += 1
                continue
            if winners[i] > nums[j]:
                j += 1
            else: 
                i += 1
        
        if card_idx == len(card_counts):
            card_counts.append(1)
        
        count = card_counts[card_idx]

        for i in range(card_idx + 1, card_idx+win_count+1):
            if i == len(card_counts):
                card_counts.append(count+1)
            else:
                card_counts[i] += count

        card_idx += 1

print(sum(card_counts))

File: day4/second/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"slices"
	"strconv"
	"strings"
)

func main() {
	input, err := os.Open("../input.txt")
	if err != nil {
		os.Exit(1)
	}
	scanner := bufio.NewScanner(input)

	cardCounts := []int{}
	sum := 0

	cardIdx := 0
	for scanner.Scan() {
		card := scanner.Text()
		parts := strings.Split(card, "|")
		winners := atoi(strings.Fields(strings.TrimSpace(parts[0][strings.Index(parts[0], ":")+1:])))
		nums := atoi(strings.Fields(strings.TrimSpace(parts[1])))
		slices.Sort(winners)
		slices.Sort(nums)

		winCount := 0
		for i, j := 0, 0; i < len(winners) && j < len(nums); {
			if winners[i] == nums[j] {
				winCount++
				i++
				j++
				continue
			}
			if winners[i] > nums[j] {
				j++
			} else {
				i++
			}
		}

		if cardIdx == len(cardCounts) {
			cardCounts = append(cardCounts, 1)
		}

		count := cardCounts[cardIdx]

		for i := cardIdx + 1; i <= cardIdx+winCount; i++ {
			if i == len(cardCounts) {
				cardCounts = append(cardCounts, count+1)
			} else {
				cardCounts[i] += count
			}
		}
		cardIdx++
	}
	for _, val := range cardCounts {
		sum += val
	}
	fmt.Println(sum)
}

func atoi(slice []string) []int {
	intSlice := []int{}
	for _, e := range slice {
		num, _ := strconv.Atoi(e)
		intSlice = append(intSlice, num)
	}
	return intSlice
}
